manim/utils/decorators.py

```diff
@@ -0,0 +1,454 @@
+        )
+        def foo():
+            pass
+
+        foo()
+        # WARNING  The function foo has been deprecated since v0.2 and is expected to be deleted after v0.4. Use bar instead. It is cooler.
+
+    You may also use dates instead of versions::
+
+        @deprecated(since="05/01/2021", until="06/01/2021")
+        def foo():
+            pass
+
+        foo()
+        # WARNING  The function foo has been deprecated since 05/01/2021 and is expected to be deleted after 06/01/2021.
+
+    """
+    # If used as factory:
+    if func is None:
+        return lambda func: deprecated(func, since, until, replacement, message)
+
+    what, name = __get_callable_info(func)
+
+    def warning_msg(for_docs: bool = False) -> str:
+        """Generate the deprecation warning message.
+
+        Parameters
+        ----------
+        for_docs
+            Whether or not to format the message for use in documentation.
+
+        Returns
+        -------
+        str
+            The deprecation message.
+        """
+        message_ = message
+        if replacement is not None:
+            replacement_ = replacement
+            if for_docs:
+                mapper = {"class": "class", "method": "meth", "function": "func"}
+                replacement_ = f":{mapper[what]}:`~.{replacement}`"
+            message_ = f"Use {replacement_} instead. {message}"
+        deprecated = __deprecation_text_component(since, until, message_)
+        return f"The {what} {name} has been {deprecated}"
+
+    def deprecate_docs(func: Callable):
+        """Adjust doc string to indicate the deprecation.
+
+        Parameters
+        ----------
+        func
+            The callable whose docstring to adjust.
+        """
+        warning = warning_msg(True)
+        func.__doc__ = f"Deprecated.\n .. warning::\n  {warning}\n{func.__doc__}"
+
+    def deprecate(func: Callable, *args, **kwargs):
+        """The actual decorator used to extend the callables behavior.
+
+        Logs a warning message.
+
+        Parameters
+        ----------
+        func
+            The callable to decorate.
+        args
+            The arguments passed to the given callable.
+        kwargs
+            The keyword arguments passed to the given callable.
+
+        Returns
+        -------
+        Any
+            The return value of the given callable when beeing passed the given
+            arguments.
+        """
+        logger.warning(warning_msg())
+        return func(*args, **kwargs)
+
+    if type(func) == type:
+        deprecate_docs(func)
+        func.__init__ = decorate(func.__init__, deprecate)
+        return func
+
+    func = decorate(func, deprecate)
+    deprecate_docs(func)
+    return func
+
+
+def deprecated_params(
+    params: Union[str, Iterable[str]] = [],
+    since: Optional[str] = None,
+    until: Optional[str] = None,
+    message: str = "",
+    redirections: "Iterable[Union[Tuple[str, str], Callable[..., dict[str, Any]]]]" = [],
```

Comment:
I believe there shouldn't be quotes here ?

Reply:
Yup.